include/onnxruntime/core/graph/graph.h

```diff
@@ -795,6 +784,77 @@ class Graph {
   /** Returns the mutable parent graph if this is a subgraph */
   Graph* MutableParentGraph() { return parent_graph_; }
 
+  /** Sets the type of a NodeArg, replacing existing type/shape if any */
+  void SetNodeArgType(NodeArg& arg, const onnx::TypeProto& type_proto);
+
+  const Node* GetProducerNode(const std::string& node_arg_name) const {
+    return GetProducerNodeImpl(*this, node_arg_name);
+  }
+
+  Node* GetMutableProducerNode(const std::string& node_arg_name) {
+    return GetProducerNodeImpl(*this, node_arg_name);
+  }
+
+  void UpdateProducerNode(const std::string& node_arg_name, NodeIndex node_index) {
+    auto iter = node_arg_to_producer_node_.find(node_arg_name);
+
+    if (iter != node_arg_to_producer_node_.end()) {
+      iter->second = node_index;
+    } else {
+      node_arg_to_producer_node_[node_arg_name] = node_index;
+    }
+  }
+
+  std::vector<const Node*> GetConsumerNodes(const std::string& node_arg_name) const {
+    return GetConsumerNodesImpl(*this, node_arg_name);
+  }
+
+  std::vector<Node*> GetMutableConsumerNodes(const std::string& node_arg_name) {
+    return GetConsumerNodesImpl(*this, node_arg_name);
+  }
+
+  void UpdateConsumerNodes(const std::string& node_arg_name, const std::vector<Node*>& nodes) {
+    auto iter = node_arg_to_consumer_nodes_.find(node_arg_name);
+    if (iter != node_arg_to_consumer_nodes_.end()) {
+      node_arg_to_consumer_nodes_.erase(node_arg_name);
+    }
+    for (Node* node : nodes) {
+      node_arg_to_consumer_nodes_[node_arg_name].insert(node->Index());
+    }
+  }
+
+  // Options to control Graph::Resolve.
+  struct ResolveOptions {
+    bool override_types = false;
+    const std::unordered_set<std::string>* initializer_names_to_preserve = nullptr;
+    bool no_proto_sync_required = false;
+  };
```

Comment:
nit: now that these are in a header it would be good to add an description saying what each option does.